### PR TITLE
IGNITE-20322 Add ability to pass an observable timestamp to an implicit transaction

### PR DESCRIPTION
**Pull request by vldpyatkov**
https://issues.apache.org/jira/browse/IGNITE-20322

### CODE REVIEW

**Comment by vldpyatkov** [comment]
Incorrect TC runs were attached.
https://ci.ignite.apache.org/buildConfiguration/ApacheIgnite3xGradle_Test_RunAllTests?branch=pull%2F2537&buildTypeTab=overview&mode=builds

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientObservationTsTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        clientTable.keyValueView().put(null, Tuple.create().set(COLUMN_KEY, 42), Tuple.create().set(COLUMN_VAL, "client value"));
+
+        Transaction tx =  client().transactions().begin(new TransactionOptions().readOnly(true));
+
+        String clientValue = clientTable.keyValueView().get(tx,  Tuple.create().set(COLUMN_KEY, 42)).value(COLUMN_VAL);
+
+        tx.commit();
+
+        Transaction srvTx =  server().transactions().begin(new TransactionOptions().readOnly(true));
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientObservationTsTest.java:50]: plz configure _idleSafeTimePropagationDuration_ before this and the same tests, for avoid potential flaky on jvm freeze

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientObservationTsTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * This test is checking correctness of observation timestamp calculation.
+ */
+public class ItThinClientObservationTsTest extends ItAbstractThinClientTest {
+
+    @Test
+    public void test() {
+        Table srvTable = server().tables().table(TABLE_NAME);
+        Table clientTable = client().tables().table(TABLE_NAME);
+
+        log.info("Tables [client={}, srv={}]", clientTable.getClass().getSimpleName(), srvTable.getClass().getSimpleName());
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientObservationTsTest.java:38]: remove or change log string

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java [hunk 3]
```diff
@@ -661,49 +661,49 @@ private void processOperation(ChannelHandlerContext ctx, ClientMessageUnpacker i
                 return ClientTableGetRequest.process(in, out, igniteTables);
 
             case ClientOp.TUPLE_UPSERT:
-                return ClientTupleUpsertRequest.process(in, out, igniteTables, resources);
+                return ClientTupleUpsertRequest.process(in, out, igniteTables, resources, igniteTransactions);
```
Comment (ascherbakoff) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java:664]: We don't need to track a client observable ts on server because it's already tracked on client for RO purposes:
`org.apache.ignite.internal.client.ReliableChannel#observableTimestamp`
This should be fixed. Instead, the commit ts should be propagated directly to client on commit.

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java [hunk 4]
```diff
@@ -135,9 +140,24 @@ private void checkEnlistPossibility() {
...
+    @Override
+    public CompletableFuture<Void> rollbackAsync() {
+        return TransactionsExceptionMapperUtil.convertToPublicFuture(
+                finish(false, nullableHybridTimestamp(NULL_HYBRID_TIMESTAMP), false),
+                TX_ROLLBACK_ERR
+        );
+    }
+
+    @Override
+    public CompletableFuture<Void> finish(boolean commit, HybridTimestamp executionTimestamp, boolean full) {
```
Comment (ascherbakoff) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java:160]: Why execution timestamp is not nullable?
This is to avoid constructing useless object.

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/IgniteTransactionsImpl.java [hunk 5]
```diff
@@ -103,8 +103,18 @@ public CompletableFuture<Transaction> beginAsync(@Nullable TransactionOptions op
         return CompletableFuture.completedFuture(begin(options));
     }
 
+    /**
+     * Begins a transaction.
+     *
+     * @param readOnly {@code true} in order to start a read-only transaction, {@code false} in order to start read-write one.
+     * @return The started transaction.
+     */
+    public InternalTransaction beginImplicit(boolean readOnly) {
```
Comment (ascherbakoff) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/IgniteTransactionsImpl.java:112]: I believe it's not OK to add implicit flag to a transaction state as the part of this PR.
This introduces a log of unrelated bloating changes and better to be done as separate PR.

**Review by ascherbakoff — CHANGES_REQUESTED** [review]

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/TransactionIds.java [hunk 6]
```diff
@@ -34,8 +37,8 @@ public class TransactionIds {
      * @param priority Transaction priority.
      * @return Transaction ID corresponding to the provided values.
      */
-    public static UUID transactionId(HybridTimestamp beginTimestamp, int nodeId, TxPriority priority) {
-        return transactionId(beginTimestamp.longValue(), nodeId, priority);
+    public static UUID transactionId(HybridTimestamp beginTimestamp, int nodeId, boolean implicit, TxPriority priority) {
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/TransactionIds.java:40]: Fortunately, at this moment we do not have to bother about compatibility, but I will try to move the flag out of transaction id.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetAllRequest.java [hunk 7]
```diff
@@ -49,6 +49,7 @@ public static CompletableFuture<Void> process(
             ClientResourceRegistry resources
     ) {
         return readTableAsync(in, tables).thenCompose(table -> {
+            // TODO: IGNITE-23603 We have to create an implicit transaction, but leave a possibility to start RO direct.
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetAllRequest.java:52]: I thought again, and currently we still have this issue only in batch get operation. 
We have to determine which transaction has to be started, RO or RW. It depends on if all keys are in one partition or not.

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/TransactionIds.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/TransactionIds.java]: I changed it to prived field in transaction object.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleDeleteAllExactRequest.java [hunk 8]
```diff
@@ -35,20 +36,22 @@ public class ClientTupleDeleteAllExactRequest {
...
             ClientMessageUnpacker in,
             ClientMessagePacker out,
             IgniteTables tables,
-            ClientResourceRegistry resources
+            ClientResourceRegistry resources,
+            IgniteTransactionsImpl igniteTransactions
     ) {
         return readTableAsync(in, tables).thenCompose(table -> {
-            var tx = readTx(in, out, resources);
+            var tx = readOrStartImplicitTx(in, out, resources, igniteTransactions, false);
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleDeleteAllExactRequest.java]: I fixed it in all places in the client handler where we are starting implicit transactions now.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java (hunk 3 above)
Comment (vldpyatkov) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java:664]: Currently, we can't do that. Probably we might think about this later.

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/IgniteTransactionsImpl.java (hunk 5 above)
Comment (vldpyatkov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/IgniteTransactionsImpl.java:112]: The possibility to start an implicit transaction in another component is a main cause of this PR. Without this, we cannot handle an observable timestamp.

**Review by denis-chudov — APPROVED** [review]

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/IgniteTransactionsImpl.java (hunk 5 above)
Comment (ascherbakoff) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/IgniteTransactionsImpl.java:112]: All beginXXX methods alredy accept HybridTimestampTracker as an argument.
To fix the issue in the ticket description it's not necessary to add `implicit `state field to a transaction.

**Review by ascherbakoff — APPROVED** [review]
